File: urbanairship-preference-center-react-native/android/src/main/java/com/urbanairship/reactnative/preferenceCenter/AirshipPreferenceCenterModule.java
/* Copyright Urban Airship and Contributors */

package com.urbanairship.reactnative.preferenceCenter;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.module.annotations.ReactModule;
import com.urbanairship.UAirship;
import com.urbanairship.preferencecenter.PreferenceCenter;
import com.urbanairship.reactnative.Event;
import com.urbanairship.reactnative.EventEmitter;
import com.urbanairship.reactnative.ReactAirshipPreferences;
import com.urbanairship.reactnative.Utils;
import com.urbanairship.reactnative.preferenceCenter.events.OpenPreferenceCenterEvent;

@ReactModule(name = AirshipPreferenceCenterModule.NAME)
public class AirshipPreferenceCenterModule extends ReactContextBaseJavaModule {

    public static final String NAME = "AirshipPreferenceCenterModule";
    public final ReactAirshipPreferences preferences;

    public AirshipPreferenceCenterModule(ReactApplicationContext reactContext) {
        super(reactContext);
        preferences = ReactAirshipPreferences.shared(reactContext);
        UAirship.shared(airship -> {
            PreferenceCenter.shared().setOpenListener(preferenceCenterId -> {
                if (preferences.isAutoLaunchPreferenceCenterEnabled(preferenceCenterId)) {
                    return false;
                } else {
                    Event event = new OpenPreferenceCenterEvent(preferenceCenterId);
                    EventEmitter.shared().sendEvent(event);
                    return true;
                }
            });
        });
    }

    @Override
    @NonNull
    public String getName() {
        return NAME;
    }

    @ReactMethod
    public void open(String preferenceCenterId) {
        if (!Utils.ensureAirshipReady()) {
            return;
        }

        PreferenceCenter.shared().open(preferenceCenterId);
    }

    @ReactMethod
    public void getConfiguration(String preferenceCenterId, final Promise promise) {
        if (!Utils.ensureAirshipReady(promise)) {
            return;
        }

        PreferenceCenter.shared().getJsonConfig(preferenceCenterId).addResultCallback(result -> {
            if (result == null) {
                promise.reject(new Exception("Failed to get preference center configuration."));
                return;
            }

            promise.resolve(Utils.convertJsonValue(result));
        });
    }

    @ReactMethod
    public void setUseCustomPreferenceCenterUi(boolean useCustomUI, String preferenceID) {
        preferences.setAutoLaunchPreferenceCenter(preferenceID, !useCustomUI);
    }
}

File: urbanairship-preference-center-react-native/android/src/main/java/com/urbanairship/reactnative/preferenceCenter/AirshipPreferenceCenterModule.kt
/* Copyright Urban Airship and Contributors */

package com.urbanairship.reactnative.preferenceCenter

import android.annotation.SuppressLint
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.module.annotations.ReactModule
import com.urbanairship.PendingResult
import com.urbanairship.UAirship
import com.urbanairship.json.JsonValue
import com.urbanairship.preferencecenter.PreferenceCenter
import com.urbanairship.preferencecenter.PreferenceCenter.OnOpenListener
import com.urbanairship.reactive.Observable
import com.urbanairship.reactive.Subscriber
import com.urbanairship.reactnative.Event
import com.urbanairship.reactnative.EventEmitter
import com.urbanairship.reactnative.ReactAirshipPreferences
import com.urbanairship.reactnative.Utils
import com.urbanairship.reactnative.preferenceCenter.events.OpenPreferenceCenterEvent
import com.urbanairship.remotedata.RemoteDataPayload

@ReactModule(name = AirshipPreferenceCenterModule.NAME)
class AirshipPreferenceCenterModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {

    private val preferences: ReactAirshipPreferences by lazy { ReactAirshipPreferences.shared(reactContext) }

    init {
        UAirship.shared {
            PreferenceCenter.shared().openListener = OnOpenListener { preferenceCenterId: String ->
                    if (preferences.isAutoLaunchPreferenceCenterEnabled(preferenceCenterId)) {
                        return@OnOpenListener false
                    } else {
                        val event: Event = OpenPreferenceCenterEvent(preferenceCenterId)
                        EventEmitter.shared().sendEvent(event)
                        return@OnOpenListener true
                    }
                }
        }
    }

    override fun getName(): String {
        return NAME
    }

    @ReactMethod
    fun open(preferenceCenterId: String) {
        if (!Utils.ensureAirshipReady()) {
            return
        }
        PreferenceCenter.shared().open(preferenceCenterId)
    }

    @ReactMethod
    fun getConfiguration(preferenceCenterId: String, promise: Promise) {
        if (!Utils.ensureAirshipReady(promise)) {
            return
        }

        PreferenceCenter.shared().getJsonConfig(preferenceCenterId).addResultCallback { result: JsonValue? ->
            if (result == null) {
                promise.reject(Exception("Failed to get preference center configuration."))
                return@addResultCallback
            }
            promise.resolve(Utils.convertJsonValue(result))
        }
    }

    @ReactMethod
    fun setUseCustomPreferenceCenterUi(useCustomUI: Boolean, preferenceID: String) {
        preferences.setAutoLaunchPreferenceCenter(preferenceID, !useCustomUI)
    }

    companion object {
        const val NAME = "AirshipPreferenceCenterModule"
    }
}
